src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -1565,24 +1565,28 @@ public void Dispose()
 			using (new SessionIdLoggingContext(SessionId))
 			{
 				log.Debug(string.Format("[session-id={0}] running ISession.Dispose()", SessionId));
-				TransactionContext?.Wait();
-				if (TransactionContext != null && TransactionContext.CanFlushOnSystemTransactionCompleted)
+				var transactionContext = TransactionContext;
+				if (transactionContext != null)
```

Comment:
As you have seen by yourself, you are creating a bug here. The `Wait` must be donne before the `null` check, for protecting against races. Once the `Wait` done, no more races are possible (unless another bug kicks in), so we can work on the property directly. But since during the `Wait`, it could have been nullified, we have to check it.

And we must not store a reference of the property before the `Wait`, or we create other bugs.